### PR TITLE
Add networks to init message

### PR DESCRIPTION
**Pull request by t-bast**
Implement https://github.com/lightningnetwork/lightning-rfc/pull/682

This is a good opportunity to setup the infrastructure for a TLV stream in the Init message, which we can leverage for optional, un-speced features.

### CODE REVIEW

File: eclair-core/src/main/scala/fr/acinq/eclair/wire/LightningMessageTypes.scala [hunk 1]
```diff
@@ -45,7 +45,9 @@ sealed trait HasChainHash extends LightningMessage { def chainHash: ByteVector32
 sealed trait UpdateMessage extends HtlcMessage // <- not in the spec
 // @formatter:on
 
-case class Init(features: ByteVector) extends SetupMessage
+case class Init(features: ByteVector, tlvs: TlvStream[InitTlv] = TlvStream.empty) extends SetupMessage {
+  val networks = tlvs.get[InitTlv.Networks].map(_.chainHashes).getOrElse(Nil)
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/wire/LightningMessageTypes.scala]: In theory this should be an `Option[List]` imo, but in that particular case it makes sense to equal `Nil` and `None`.

File: eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala [hunk 2]
```diff
@@ -101,7 +101,7 @@ class Peer(val nodeParams: NodeParams, remoteNodeId: PublicKey, authenticator: A
       transport ! TransportHandler.Listener(self)
       context watch transport
       val localInit = nodeParams.overrideFeatures.get(remoteNodeId) match {
-        case Some(f) => wire.Init(f)
+        case Some(f) => wire.Init(f, TlvStream(InitTlv.Networks(nodeParams.chainHash :: Nil)))
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala]: We could factor the definition of `TlvStream(InitTlv.Networks(nodeParams.chainHash :: Nil))`.

Actually I think the pattern matching should only return `features` and we should build the `Init` after.

File: eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala [hunk 3]
```diff
@@ -148,9 +148,19 @@ class Peer(val nodeParams: NodeParams, remoteNodeId: PublicKey, authenticator: A
     case Event(remoteInit: wire.Init, d: InitializingData) =>
       d.transport ! TransportHandler.ReadAck(remoteInit)
 
-      log.info(s"peer is using features=${remoteInit.features.toBin}")
+      log.info(s"peer is using features=${remoteInit.features.toBin}, network=${remoteInit.networks}")
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala]:
```suggestion
      log.info(s"peer is using features=${remoteInit.features.toBin}, networks=${remoteInit.networks.mkString(",")}")
```

File: eclair-core/src/test/scala/fr/acinq/eclair/wire/LightningMessageCodecsSpec.scala [hunk 4]
```diff
@@ -44,21 +44,36 @@ class LightningMessageCodecsSpec extends FunSuite {
...
+    val chainHash2 = ByteVector32(hex"0202020202020202020202020202020202020202020202020202020202020202")
     val testCases = Seq(
-      (hex"0000 0000", hex"", hex"0000 0000"), // no features
-      (hex"0000 0002088a", hex"088a", hex"0000 0002088a"), // no global features
-      (hex"00020200 0000", hex"0200", hex"0000 00020200"), // no local features
-      (hex"00020200 0002088a", hex"0a8a", hex"0000 00020a8a"), // local and global - no conflict - same size
-      (hex"00020200 0003020002", hex"020202", hex"0000 0003020202"), // local and global - no conflict - different sizes
-      (hex"00020a02 0002088a", hex"0a8a", hex"0000 00020a8a"), // local and global - conflict - same size
-      (hex"00022200 000302aaa2", hex"02aaa2", hex"0000 000302aaa2") // local and global - conflict - different sizes
+      (hex"0000 0000", hex"", Nil, true, None), // no features
```
Comment (pm47) [inline eclair-core/src/test/scala/fr/acinq/eclair/wire/LightningMessageCodecsSpec.scala]: With that many elements in a tuple a case class could come in handy.

File: eclair-core/src/main/scala/fr/acinq/eclair/wire/LightningMessageTypes.scala (hunk 1 above)
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/wire/LightningMessageTypes.scala]: Agreed, that was my reasoning too.

File: eclair-core/src/test/scala/fr/acinq/eclair/wire/LightningMessageCodecsSpec.scala (hunk 4 above)
Comment (t-bast) [reply eclair-core/src/test/scala/fr/acinq/eclair/wire/LightningMessageCodecsSpec.scala]: :+1:

**Pull request merged**